### PR TITLE
Dockerize CI + Release builds

### PR DESCRIPTION
**Pull request by powderluv**
Gets both CI and Release builds integrated in one workflow.

Tested the callout with in-tree / out-of-tree and torch-mlir
release packages

TODO: add the correct CMake commands in the functions.
Mount ccache and pip cache as required

Out to build all the CI and Release builds in one go:

```
anush@denali ~/github/torch-mlir % TM_PACKAGES="out-of-tree in-tree torch-mlir" TM_TORCH_SRC=ON TM_PYTHON_VERSIONS="cp310-cp310"  ./build_tools/python_deploy/build_linux_packages.sh                                   
                                                                                                                                                                                                                     
Setting torch-mlir Python Package version to: 0.0.1                                                                                                                                                                  
Running on host                                                                                                                                                                                                      
Launching docker image stellaraccident/manylinux2014_x86_64-bazel-5.1.0:latest                                                                                                                                       
Outputting to /home/anush/github/torch-mlir/build_tools/python_deploy/wheelhouse                                                                                                                                     
Setting torch-mlir Python Package version to: 0.0.1                                                                                                                                                                  
Running in docker                                                                                                                                                                                                    
Using python versions: cp310-cp310                                                                                                                                                                                   
******************** BUILDING PACKAGE out-of-tree ********************                                                                                                                                               
:::: Python version Python 3.10.4                                                                                                                                                                                    
:::: Clean build dir out-of-tree ecp310-cp310                                                                                                                                                                        
:::: Build out-of-tree Torch from source: ON                                                                                                                                                                         
:::: Test out-of-tree                                                                                                                                                                                                
******************** BUILDING PACKAGE in-tree ********************                                                                                                                                                   
:::: Python version Python 3.10.4                                                                                                                                                                                    
:::: Clean build dir in-tree ecp310-cp310                                                                                                                                                                            
:::: Build in-tree Torch from source: ON                                                                                                                                                                             
:::: Test in-tree                                                                                                                                                                                                    
******************** BUILDING PACKAGE torch-mlir ********************                                                                                                                                                
:::: Python version Python 3.10.4                                                                                                                                                                                    
:::: Clean wheels torch_mlir cp310-cp310                                                                                                                                                                             
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/nightly/cpu                                                                                                                            
Looking in links: https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html                                                                                                                                    
Collecting torch                                                                                                                                                                                                     
  Downloading https://download.pytorch.org/whl/nightly/cpu/torch-1.13.0.dev20220816%2Bcpu-cp310-cp310-linux_x86_64.whl (195.1 MB)                                                                                    
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 195.1/195.1 MB 9.9 MB/s eta 0:00:00                                                                                                                                     
Collecting numpy                                                                                                                     
....
  adding 'torch_mlir-0.0.1.dist-info/WHEEL'
  adding 'torch_mlir-0.0.1.dist-info/top_level.txt'
  adding 'torch_mlir-0.0.1.dist-info/RECORD'
  removing build/bdist.linux-x86_64/wheel
  Building wheel for torch-mlir (pyproject.toml): finished with status 'done'
  Created wheel for torch-mlir: filename=torch_mlir-0.0.1-cp310-cp310-linux_x86_64.whl size=35159325 sha256=3f00551a9f426b8f58457a1306a143ac9f66865113dbaf2050b082e2ae543654
  Stored in directory: /root/.cache/pip/wheels/b9/1a/c9/fd0f4b77f13d9149c3c25bd33ce4a61807bcc222a8b04a5ea9
Successfully built torch-mlir
WARNING: You are using pip version 22.0.4; however, version 22.2.2 is available.
You should consider upgrading via the '/opt/python/cp310-cp310/bin/python -m pip install --upgrade pip' command.


```


If we want to use Ubuntu 22.04 for the CI:

```
anush@denali ~/github/torch-mlir % TM_PACKAGES="out-of-tree in-tree" TM_TORCH_SRC=ON TM_PYTHON_VERSIONS="cp310-cp310" TM_DOCKER_IMAGE="ubuntu:22.04" ./build_tools/python_deploy/build_linux_packages.sh

Setting torch-mlir Python Package version to: 0.0.1
Running on host
Launching docker image ubuntu:22.04
Outputting to /home/anush/github/torch-mlir/build_tools/python_deploy/wheelhouse
Unable to find image 'ubuntu:22.04' locally
22.04: Pulling from library/ubuntu
d19f32bd9e41: Pull complete 
Digest: sha256:34fea4f31bf187bc915536831fd0afc9d214755bf700b5cdb1336c82516d154e
Status: Downloaded newer image for ubuntu:22.04
```

### CODE REVIEW

**Comment by powderluv** [comment]
Gets both CI and Release builds integrated in one workflow.
Mount ccache and pip cache as required for fast iterative builds
Current Release docker builds still run with root perms, fix it
in the future to run as the same user.

There may be some corner cases left especially when switching
build types etc.

**Docker build TEST plan:**

tl;dr:
**Build everything: Releases (Python 3.8, 3.9, 3.10) and CIs.**
`  TM_PACKAGES="torch-mlir out-of-tree in-tree"
  2.57s user 2.49s system 0% cpu 30:33.11 total `

**Out of Tree + PyTorch binaries:**

  Fresh build (purged cache):
 `   TM_PACKAGES="out-of-tree"
    0.47s user 0.51s system 0% cpu 5:24.99 total`

  Incremental with ccache:
`    TM_PACKAGES="out-of-tree"
    0.09s user 0.08s system 0% cpu 34.817 total`

**Out of Tree + PyTorch from source**

  Incremental
 `   TM_PACKAGES="out-of-tree" TM_USE_PYTORCH_BINARY=OFF
    1.58s user 1.81s system 2% cpu 1:59.61 total`

**In-Tree + PyTorch binaries:**

  Fresh build and tests: (purge ccache)
`  TM_PACKAGES="in-tree"
  0.53s user 0.49s system 0% cpu 6:23.35 total`

  Fresh build/ but with prior ccache
`  TM_PACKAGES="in-tree"
  0.45s user 0.66s system 0% cpu 3:57.47 total`

  Incremental in-tree with all tests and regression tests
`  TM_PACKAGES="in-tree"
  0.16s user 0.09s system 0% cpu 2:18.52 total`

**In-Tree + PyTorch from source**

  Fresh build and tests: (purge ccache)
`  TM_PACKAGES="in-tree" TM_USE_PYTORCH_BINARY=OFF
  2.03s user 2.28s system 0% cpu 11:11.86 total`

  Fresh build/ but with prior ccache
`  TM_PACKAGES="in-tree" TM_USE_PYTORCH_BINARY=OFF
  1.58s user 1.88s system 1% cpu 4:53.15 total`

  Incremental in-tree with all tests and regression tests
`  TM_PACKAGES="in-tree" TM_USE_PYTORCH_BINARY=OFF
  1.09s user 1.10s system 1% cpu 3:29.84 total`

  Incremental without tests
`  TM_PACKAGES="in-tree" TM_USE_PYTORCH_BINARY=OFF TM_SKIP_TESTS=ON
  1.52s user 1.42s system 3% cpu 1:15.82 total`

**In-tree+out-of-tree + Pytorch Binaries**
`  TM_PACKAGES="out-of-tree in-tree"
  0.25s user 0.18s system 0% cpu 3:01.91 total`

To clear all artifacts:
`rm -rf build build_oot llvm-build libtorch docker_venv
externals/pytorch/build .ccache`

**Comment by silvasean** [comment]
This seems reasonable to me. @sjain-stanford ?

**Review by sjain-stanford — COMMENTED** [review]
Thanks @powderluv . This looks great. I will give the "local" build+test flow a try later today (very excited!). The main request I have is - since we set out to "dockerize CI" - it'd be good to also see GHA CI workflows updated to use these docker flows. This will validate the requirements fully, and ensure any cache issues or other GHA issues can be addressed alongside this PR.

**Comment by powderluv** [comment]
Happy to add the GHA pieces in the follow-on but wanted to get the base functionality in first so we don't have a mega commit and easy to revert just GHA if something goes haywire

File: docs/development.md [hunk 1]
```diff
@@ -22,6 +22,106 @@ python -m pip install --upgrade pip
...
+flow and we would like your feedback on how it works for you and please
+feel free to file any feedback or issues.
+
+Install [Docker Engine](https://docs.docker.com/engine/install/ubuntu/). You don't need Docker Desktop.
+
+You have three types of builds selectable with the Environment Variable `TM_PACKAGES`:`torch-mlir` the
+Release build, `out-of-tree` where torch-mlir is build with a pre-built MLIR and `in-tree` where torch-mlir
+is built as part of the LLVM project along with MLIR.
+
+We mount a ccache and pip cache inside the docker container to speed up iterative builds. Iterative
```
Comment (sjain-stanford) [inline docs/development.md:37]: Curious, where and how is this pip cache configured? I only saw the ccache directory mounted within the container.

File: build_tools/python_deploy/build_linux_packages.sh [hunk 2]
```diff
@@ -97,6 +165,122 @@ function run_in_docker() {
...
+      -DPython3_EXECUTABLE="$(which python3)" \
+      /main_checkout/torch-mlir
+  cmake --build /main_checkout/torch-mlir/build_oot
+  ccache -s
+}
+
+function test_out_of_tree() {
+  echo ":::: Test out-of-tree"
+  cmake --build /main_checkout/torch-mlir/build_oot --target check-torch-mlir-all
+  #TODO: Add regression tests
```
Comment (sjain-stanford) [inline build_tools/python_deploy/build_linux_packages.sh:273]: Any particular reason to leave this as a TODO and not using the integration tests similar to the in-tree case?

File: build_tools/python_deploy/build_linux_packages.sh [hunk 3]
```diff
@@ -97,6 +165,122 @@ function run_in_docker() {
...
+  local python_version="$1"
+  echo ":::: Setting up VENV with Python: $python_version"
+  python3 -m venv /main_checkout/torch-mlir/docker_venv
+  source /main_checkout/torch-mlir/docker_venv/bin/activate
+
+  echo ":::: pip installing dependencies"
+  python3 -m pip install -r /main_checkout/torch-mlir/externals/llvm-project/mlir/python/requirements.txt
+  python3 -m pip install -r /main_checkout/torch-mlir/requirements.txt
+
+}
```
Comment (sjain-stanford) [inline build_tools/python_deploy/build_linux_packages.sh:226]: Am I understanding correct that these are runtime setup steps and hence the cost is incurred for every invocation of `build_linux_packages.sh`? My worry is paying this cost for every local re-run could be avoided if these were folded into the docker build step.

File: build_tools/python_deploy/build_linux_packages.sh (hunk 2 above)
Comment (powderluv) [reply build_tools/python_deploy/build_linux_packages.sh:273]: We discussed this at the meeting this morning. @silvasean suggested just a "smoke test" is good for OOT. Will remove the TODO.

File: docs/development.md (hunk 1 above)
Comment (powderluv) [reply docs/development.md:37]: Unlike the release build that still runs as root:root I run the CI as the user. So when we mount the home directory you get the pip cache too.

File: build_tools/python_deploy/build_linux_packages.sh (hunk 3 above)
Comment (powderluv) [reply build_tools/python_deploy/build_linux_packages.sh:226]: Since it is pip cached you breeze through this in seconds (see measured runtimes).  
```
:::: pip installing dependencies                                                                                                      
Requirement already satisfied: numpy in /main_checkout/torch-mlir/docker_venv/lib/python3.10/site-packages (from -r /main_checkout/torch-mlir/externals/llvm-project/mlir/python/requirements.txt (line 1)) (1.23.2)
Requirement already satisfied: pybind11>=2.8.0 in /main_checkout/torch-mlir/docker_venv/lib/python3.10/site-packages (from -r /main_checkout/torch-mlir/externals/llvm-project/mlir/python/requirements.txt (line 2)) (2.10.0)
....
```

My entire OOT + In-tree build takes 3mins if I touch one file. 
```
TM_PACKAGES="out-of-tree in-tree" TM_PYTHON_VERSIONS="cp310-cp310"   0.20s user 0.18s system 0% cpu 3:16.42 total
```

This also takes care of cases when someone makes a change to requirements.txt.  We don't want end users to have to `docker build` it is an unnecessary burden.

File: build_tools/python_deploy/build_linux_packages.sh (hunk 2 above)
Comment (powderluv) [reply build_tools/python_deploy/build_linux_packages.sh:273]: updated in latest PR

**Comment by powderluv** [comment]
I have also added the GHA workflows now.

File: build_tools/python_deploy/build_linux_packages.sh [hunk 4]
```diff
@@ -16,77 +16,145 @@
...
           #run_audit_wheel torch_mlir "$python_version"
+          clean_build torch_mlir "$python_version"
+          ;;
+        out-of-tree)
+          #clean_build "$package" "$python_version"
+          setup_venv "$python_version"
+          build_out_of_tree "$TM_USE_PYTORCH_BINARY" "$python_version"
+          if [ "${TM_SKIP_TESTS}" == "OFF" ]; then
+            #Tests always run with the installed PyTorch to simulate end users
+            #setup_venv "$python_version"
```
Comment (sjain-stanford) [inline build_tools/python_deploy/build_linux_packages.sh:145]: nit: dead code?

File: build_tools/python_deploy/build_linux_packages.sh [hunk 5]
```diff
@@ -16,77 +16,145 @@
...
+        echo "Defaulting to $python_dir (expected for CI builds)"
       fi
       export PATH=$python_dir/bin:$orig_path
-      echo ":::: Python version $(python --version)"
+      echo ":::: Python version $(python3 --version)"
       case "$package" in
         torch-mlir)
           clean_wheels torch_mlir "$python_version"
           build_torch_mlir
           #run_audit_wheel torch_mlir "$python_version"
```
Comment (sjain-stanford) [inline build_tools/python_deploy/build_linux_packages.sh:136]: nit: dead code?

File: build_tools/python_deploy/build_linux_packages.sh [hunk 6]
```diff
@@ -16,77 +16,145 @@
...
+          clean_build torch_mlir "$python_version"
+          ;;
+        out-of-tree)
+          #clean_build "$package" "$python_version"
```
Comment (sjain-stanford) [inline build_tools/python_deploy/build_linux_packages.sh:140]: nit: dead code?

File: build_tools/python_deploy/build_linux_packages.sh [hunk 7]
```diff
@@ -16,77 +16,145 @@
...
+          setup_venv "$python_version"
+          build_out_of_tree "$TM_USE_PYTORCH_BINARY" "$python_version"
+          if [ "${TM_SKIP_TESTS}" == "OFF" ]; then
+            #Tests always run with the installed PyTorch to simulate end users
+            #setup_venv "$python_version"
+            test_out_of_tree
+	        fi
+          ;;
+        in-tree)
+          #clean_build "$package" "$python_version"
```
Comment (sjain-stanford) [inline build_tools/python_deploy/build_linux_packages.sh:150]: nit: dead code?

File: build_tools/python_deploy/build_linux_packages.sh [hunk 8]
```diff
@@ -16,77 +16,145 @@
...
+          setup_venv "$python_version"
+          build_in_tree "$TM_USE_PYTORCH_BINARY" "$python_version"
+          if [ "${TM_SKIP_TESTS}" == "OFF" ]; then
+            #Tests always run with the installed PyTorch to simulate end users
+            #setup_venv "$python_version"
```
Comment (sjain-stanford) [inline build_tools/python_deploy/build_linux_packages.sh:155]: nit: dead code?

File: docs/development.md [hunk 9]
```diff
@@ -22,6 +22,106 @@ python -m pip install --upgrade pip
...
+
+### Mimicing CI+Release builds
+
+If you wanted to build all the CIs and the Release builds (just with Python 3.10 since most other Python builds are redundant)
+
+```shell
+TM_PACKAGES="torch-mlir out-of-tree in-tree" TM_PYTHON_VERSIONS="cp310-cp310" ./build_tools/python_deploy/build_linux_packages.sh
+```
+
+Note: The Release docker still runs as root so it may generate some files owned by root:root. We hope to move it to run as a user in the future.
```
Comment (sjain-stanford) [inline docs/development.md:71]: It may be useful to have a section in the README that just shows the commands for running CI locally (without combining with release builds).

**Review by sjain-stanford — APPROVED** [review]
LG(reat)TM. From my local testing, I can confirm that re-runs are blazing fast (utilize pip cache)! Left some minor comments to get this going. Also, it seems #1313 doesn't have GHA workflows yet - I'm showing this same PR replicated there, could you PTAL?. Again, thanks for working on the follow-on commit to validate the GHA workflows as well.

File: build_tools/python_deploy/build_linux_packages.sh [hunk 10]
```diff
@@ -97,6 +165,122 @@ function run_in_docker() {
...
+      -DPython3_EXECUTABLE="$(which python3)" \
+      /main_checkout/torch-mlir
+  cmake --build /main_checkout/torch-mlir/build_oot
+  ccache -s
+}
+
+function test_out_of_tree() {
+  echo ":::: Test out-of-tree"
+  cmake --build /main_checkout/torch-mlir/build_oot --target check-torch-mlir-all
+  #TODO: Add regression tests
```
[the remark this answers is not in the document]
Comment (powderluv) [reply build_tools/python_deploy/build_linux_packages.sh:273]: will do. somehow got re-inserted in git patch shuffle.

File: build_tools/python_deploy/build_linux_packages.sh [hunk 11]
```diff
@@ -16,77 +16,145 @@
...
+      ;;
+  esac
+  echo "Launching docker image ${TM_CURRENT_DOCKER_IMAGE} with UID:${USERID} GID:${GROUPID}"
   docker run --rm \
     -v "${repo_root}:/main_checkout/torch-mlir" \
-    -v "${output_dir}:/wheelhouse" \
+    -v "${TM_OUTPUT_DIR}:/wheelhouse" \
+    -v "${HOME}:/home/${USER}" \
+    --user ${USERID}:${GROUPID} \
+    --workdir="/home/$USER" \
```
[the remark this answers is not in the document]
Comment (powderluv) [reply build_tools/python_deploy/build_linux_packages.sh:97]: Yeah I had to override the default `/main_checkout/torch-mlir` to the current user $USER and $USER isn't known ahead.

File: build_tools/python_deploy/build_linux_packages.sh [hunk 12]
```diff
@@ -16,77 +16,145 @@
...
+    --volume="/etc/group:/etc/group:ro" \
+    --volume="/etc/passwd:/etc/passwd:ro" \
+    --volume="/etc/shadow:/etc/shadow:ro" \
+    --ipc=host \
+    --ulimit nofile=32768:32768 \
```
[the remark this answers is not in the document]
Comment (powderluv) [reply build_tools/python_deploy/build_linux_packages.sh:102]: good questions. You need the unlimt since the default number of open files 1024 is not sufficient for our e2e tests. You need to use --ipc=host to be able to use the host sizes etc.

File: build_tools/python_deploy/build_linux_packages.sh (hunk 4 above)
Comment (powderluv) [reply build_tools/python_deploy/build_linux_packages.sh:145]: removed in next patch.

File: build_tools/python_deploy/build_linux_packages.sh (hunk 5 above)
Comment (powderluv) [reply build_tools/python_deploy/build_linux_packages.sh:136]: I would leave this since (it already exists). We need to uncomment and test our Release builds before pushing to PyPI

File: build_tools/python_deploy/build_linux_packages.sh (hunk 6 above)
Comment (powderluv) [reply build_tools/python_deploy/build_linux_packages.sh:140]: removed in next patch.

File: build_tools/python_deploy/build_linux_packages.sh (hunk 7 above)
Comment (powderluv) [reply build_tools/python_deploy/build_linux_packages.sh:150]: removed in next patch.

File: docs/development.md (hunk 9 above)
Comment (powderluv) [reply docs/development.md]: done

File: build_tools/docker/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,54 @@
+ARG BASE_IMG=ubuntu:22.04
```
[the remark this answers is not in the document]
Comment (powderluv) [reply build_tools/docker/Dockerfile:1]: Yes. The Dockerfile is checked in too and I add some notes on how to build it locally if you want to.

**Comment by powderluv** [comment]
Yeah I noticed a timeout yesterday too and a rerun ran faster. There was no functional change for the release build but if you noticed anything that got added that could affect it please let me know

**Comment by powderluv** [comment]
Actually looking at the code somemore we did change the docker settings to use `--ipc=host` and bumped the ulimit so the tests can pass. So it is possible the host this VM is running on is slow w.r.t the IPC etc. 

Opened https://github.com/llvm/torch-mlir/issues/1322 to investigate
